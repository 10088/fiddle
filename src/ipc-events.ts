export enum IpcEvents {
  MAIN_WINDOW_READY_TO_SHOW = 'MAIN_WINDOW_READY_TO_SHOW',
  OPEN_SETTINGS = 'OPEN_SETTINGS',
  LOAD_GIST_REQUEST = 'LOAD_GIST_REQUEST',
  RUN_FIDDLE = 'RUN_FIDDLE',
  MONACO_EXECUTE_COMMAND = 'MONACO_EXECUTE_COMMAND',
  FS_NEW_FIDDLE = 'FS_NEW_FIDDLE',
  FS_OPEN_FIDDLE = 'FS_OPEN_FIDDLE',
  FS_SAVE_FIDDLE = 'FS_SAVE_FIDDLE',
  FS_SAVE_FIDDLE_AS = 'FS_SAVE_FIDDLE_AS',
  FS_SAVE_FIDDLE_GIST = 'FS_SAVE_FIDDLE_GIST',
  FS_SAVE_FIDDLE_DIALOG = 'FS_SAVE_FIDDLE_DIALOG',
  FS_SAVE_FIDDLE_ERROR = 'FS_SAVE_FIDDLE_ERROR',
}

export const ipcMainEvents = [
  IpcEvents.MAIN_WINDOW_READY_TO_SHOW,
  IpcEvents.FS_SAVE_FIDDLE_DIALOG,
  IpcEvents.FS_SAVE_FIDDLE
];

export const ipcRendererEvents = [
  IpcEvents.OPEN_SETTINGS,
  IpcEvents.LOAD_GIST_REQUEST,
  IpcEvents.RUN_FIDDLE,
  IpcEvents.MONACO_EXECUTE_COMMAND,
  IpcEvents.FS_NEW_FIDDLE,
  IpcEvents.FS_OPEN_FIDDLE,
  IpcEvents.FS_SAVE_FIDDLE,
  IpcEvents.FS_SAVE_FIDDLE_AS,
  IpcEvents.FS_SAVE_FIDDLE_GIST,
];
