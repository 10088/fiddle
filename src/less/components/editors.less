.mosaic-tile {
  animation: fadein 0.15s;
}

.mosaic-window-toolbar  {
  & > div {
    display: flex;
    justify-content: space-between;
    width: 100%;
    padding: 0 3px 0 29px;

    .mosaic-controls {
      transform: scale(0.7);
    }

    h5 {
      margin: 4px 0 0 0;
    }

    button {
      margin-left: 8px;
    }
  }
}

@keyframes fadein {
  from { opacity: 0; }
  to { opacity: 1; }
}
