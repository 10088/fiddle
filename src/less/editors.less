.mosaic-tile {
  animation: fadein 0.4s;
}

@keyframes fadein {
  from { opacity: 0; }
  to { opacity: 1; }
}
