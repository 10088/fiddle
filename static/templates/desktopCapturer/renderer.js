const { desktopCapturer } = require('electron')

// The following example shows how to capture video from
// the screen. It also grabs each window, so you could
// just grab video from a single window.
//
desktopCapturer.getSources({ types: ['window', 'screen'] }, (error, sources) => {
  if (error) throw error
  for (let i = 0; i < sources.length; ++i) {
    console.log(sources[i])

    if (sources[i].name.startsWith('Screen')) {
      navigator.mediaDevices.getUserMedia({
        audio: false,
        video: {
          mandatory: {
            chromeMediaSource: 'desktop',
            chromeMediaSourceId: sources[i].id,
            minWidth: 1280,
            maxWidth: 1280,
            minHeight: 720,
            maxHeight: 720
          }
        }
      })
      .then((stream) => handleStream(stream))
      .catch((e) => handleError(e))
      return
    }
  }
})

function handleStream (stream) {
  const video = document.querySelector('video')
  video.srcObject = stream
  video.onloadedmetadata = (e) => video.play()
}

function handleError (e) {
  console.log(e)
}