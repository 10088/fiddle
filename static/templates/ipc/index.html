<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>IPC Example</title>
</head>
<body>
  <h1>IPC Example</h1>
  <p>Open the developer tools to see the IPC in action.</p>
</body>
</html>
