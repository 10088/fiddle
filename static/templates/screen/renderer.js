const { screen } = require('electron')

const screens = screen.getAllDisplays()
const div = document.querySelector('div')

div.innerText = JSON.stringify(screens);
