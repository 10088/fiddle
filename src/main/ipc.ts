import { EventEmitter } from 'events';
import { ipcMain } from 'electron';
import { ipcMainEvents } from '../ipc-events';

/**
 * The main purpose of this class is to be the central
 * gathering place for IPC calls the main process sends
 * or listens to.
 *
 * @class IpcManager
 * @extends {EventEmitter}
 */
export class IpcMainManager extends EventEmitter {
  constructor() {
    super();

    ipcMainEvents.forEach((name) => {
      ipcMain.on(name, (...args) => this.emit(name, ...args));
    });
  }
}

export const ipcMainManager = new IpcMainManager();
