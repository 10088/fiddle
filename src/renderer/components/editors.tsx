import * as MonacoType from 'monaco-editor';
import * as React from 'react';
import { Mosaic, MosaicWindow } from 'react-mosaic-component';

import { EditorId } from '../../interfaces';
import { IpcEvents } from '../../ipc-events';
import { updateEditorLayout } from '../../utils/editor-layout';
import { getFocusedEditor } from '../../utils/focused-editor';
import { getAtPath, setAtPath } from '../../utils/js-path';
import { toggleMonaco } from '../../utils/toggle-monaco';
import { ContentNames, getContent } from '../content';
import { ipcRendererManager } from '../ipc';
import { AppState } from '../state';
import { activateTheme } from '../themes';
import { Editor } from './editor';

const options: any = {
  direction: 'row',
  first: 'main',
  second: {
    direction: 'column',
    first: 'renderer',
    second: 'html',
  },
};

const defaultMonacoOptions: MonacoType.editor.IEditorOptions = {
  minimap: {
    enabled: false
  },
  wordWrap: 'on'
};

const ViewIdMosaic = Mosaic.ofType<EditorId>();
const ViewIdMosaicWindow = MosaicWindow.ofType<EditorId>();

const TITLE_MAP: Record<EditorId, string> = {
  main: 'Main Process',
  renderer: 'Renderer Process',
  html: 'HTML'
};

export interface EditorsProps {
  appState: AppState;
}

export interface EditorsState {
  monaco?: typeof MonacoType;
  isMounted?: boolean;
  monacoOptions: MonacoType.editor.IEditorOptions;
}

export class Editors extends React.Component<EditorsProps, EditorsState> {
  constructor(props: EditorsProps) {
    super(props);
    this.state = { monacoOptions: defaultMonacoOptions };
    this.loadMonaco();
  }

  public componentDidMount() {
    ipcRendererManager.on(IpcEvents.MONACO_EXECUTE_COMMAND, (_event, cmd: string) => {
      this.executeCommand(cmd);
    });

    ipcRendererManager.on(IpcEvents.FS_NEW_FIDDLE, async (_event) => {
      const { version } = this.props.appState;

      this.props.appState.setWarningDialogTexts({
        label: 'Your current fiddle is unsaved. Do you want to discard it?'
      });

      window.ElectronFiddle.app.setValues({
        html: await getContent(ContentNames.HTML, version),
        renderer: await getContent(ContentNames.RENDERER, version),
        main: await getContent(ContentNames.MAIN, version),
      });
    });

    ipcRendererManager.on(IpcEvents.MONACO_TOGGLE_OPTION, (_event, cmd: string) => {
      this.toggleEditorOption(cmd);
    });

    this.setState({ isMounted: true });
  }

  /**
   * Attempt to execute a given commandId on the currently focused editor
   *
   * @param {string} commandId
   * @memberof Editors
   */
  public executeCommand(commandId: string) {
    const editor = getFocusedEditor();

    if (editor) {
      const command = editor.getAction(commandId);

      console.log(`Editors: Trying to run ${command.id}. Supported: ${command.isSupported}`);

      if (command && command.isSupported()) {
        command.run();
      }
    }
  }

  public toggleEditorOption(path: string) {
    if (!window || !window.ElectronFiddle || !window.ElectronFiddle.editors) {
      return;
    }

    try {
      const { monacoOptions } = this.state;
      const newOptions = { ...monacoOptions };
      const currentSetting = getAtPath(path, newOptions);

      setAtPath(path, newOptions, toggleMonaco(currentSetting));

      Object.keys(window.ElectronFiddle.editors)
      .forEach((key) => {
        const editor: MonacoType.editor.IStandaloneCodeEditor | null
          = window.ElectronFiddle.editors[key];

        if (editor) {
          editor.updateOptions(newOptions);
        }
      });

      this.setState({ monacoOptions: newOptions });
    } catch (error) {
      console.warn(`Editors: Could not toggle property ${path}`, error);
    }
  }

  public render() {
    const { appState } = this.props;
    const { monaco } = this.state;

    if (!monaco) return null;

    return (
      <ViewIdMosaic
        onChange={updateEditorLayout}
        // tslint:disable-next-line:jsx-no-multiline-js
        renderTile={(id, path) => (
          <ViewIdMosaicWindow className={id} path={path} title={TITLE_MAP[id]} toolbarControls={<div />}>
            <Editor id={id} monaco={monaco} appState={appState} monoacoOptions={defaultMonacoOptions} />
          </ViewIdMosaicWindow>
        )}
        initialValue={options}
      />
    );
  }

  /**
   * Loads monaco. If it's already loaded, it'll just set it on the current state.
   * We're doing things a bit roundabout to ensure that we're not overloading the
   * mobx state with a gigantic Monaco tree.
   *
   * @private
   */
  private async loadMonaco(): Promise<void> {
    const { app } = window.ElectronFiddle;
    const loader = require('monaco-loader');
    const monaco = app.monaco || await loader();

    if (!app.monaco) {
      app.monaco = monaco;
    }

    if (!this.state || !this.state.isMounted) {
      this.state = { monaco, monacoOptions: defaultMonacoOptions };
    } else {
      this.setState({ monaco });
    }

    activateTheme(monaco, undefined, this.props.appState.theme);
  }
}
