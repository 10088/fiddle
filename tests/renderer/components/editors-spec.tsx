import { shallow } from 'enzyme';
import * as React from 'react';

import { IpcEvents } from '../../../src/ipc-events';
import { Editors } from '../../../src/renderer/components/editors';
import { ipcRendererManager } from '../../../src/renderer/ipc';
import { getFocusedEditor } from '../../../src/utils/focused-editor';

jest.mock('monaco-loader', () => jest.fn(async () => {
  return { monaco: true };
}));

jest.mock('../../../src/renderer/components/editor', () => ({
  Editor: 'Editor'
}));

jest.mock('../../../src/utils/focused-editor', () => ({
  getFocusedEditor: jest.fn()
}));

describe('Editrors component', () => {
  let store: any;
  let monaco: any;

  beforeEach(() => {
    store = {
      isTokenDialogShowing: false,
      isSettingsShowing: false,
      setWarningDialogTexts: () => ({})
    };

    monaco = {
      editor: {
        defineTheme: jest.fn()
      }
    };
  });

  it('renders', () => {
    const wrapper = shallow(<Editors appState={store} />);
    wrapper.setState({ monaco });
    expect(wrapper).toMatchSnapshot();
  });

  it('does not execute command if not supported', () => {
    const wrapper = shallow(<Editors appState={store} />);
    const instance: Editors = wrapper.instance() as any;
    const mockAction = {
      isSupported: jest.fn(() => false),
      run: jest.fn()
    };
    const mockEditor = {
      getAction: jest.fn(() => mockAction)
    };

    (getFocusedEditor as any).mockReturnValueOnce(mockEditor);

    instance.executeCommand('hello');
    expect(mockEditor.getAction).toHaveBeenCalled();
    expect(mockAction.isSupported).toHaveBeenCalled();
    expect(mockAction.run).toHaveBeenCalledTimes(0);
  });

  describe('toggleEditorOption()', () => {
    it('handles a missing ElectronFiddle global', () => {
      window.ElectronFiddle = undefined as any;

      const wrapper = shallow(<Editors appState={store} />);
      const instance: Editors = wrapper.instance() as any;

      expect(instance.toggleEditorOption('wordWrap')).toBe(false);
    });

    it('handles an error', () => {
      (window.ElectronFiddle.editors.html!.updateOptions as jest.Mock).mockImplementationOnce(
        () => {
          throw new Error('Bwap bwap');
        }
      );

      const wrapper = shallow(<Editors appState={store} />);
      const instance: Editors = wrapper.instance() as any;

      expect(instance.toggleEditorOption('wordWrap')).toBe(false);
    });

    it('updates a setting', () => {
      const wrapper = shallow(<Editors appState={store} />);
      const instance: Editors = wrapper.instance() as any;

      expect(instance.toggleEditorOption('wordWrap')).toBe(true);
      expect(window.ElectronFiddle.editors.html!.updateOptions).toHaveBeenCalledWith(
        {
          minimap: { enabled: false },
          wordWrap: 'off'
        }
      );
    });
  });

  describe('IPC commands', () => {
    it('handles an MONACO_EXECUTE_COMMAND command', () => {
      shallow(<Editors appState={store} />);

      const mockAction = {
        isSupported: jest.fn(() => true),
        run: jest.fn()
      };
      const mockEditor = {
        getAction: jest.fn(() => mockAction)
      };

      (getFocusedEditor as any).mockReturnValueOnce(mockEditor);

      ipcRendererManager.emit(IpcEvents.MONACO_EXECUTE_COMMAND, null, 'hello');
      expect(mockEditor.getAction).toHaveBeenCalled();
      expect(mockAction.isSupported).toHaveBeenCalled();
      expect(mockAction.run).toHaveBeenCalled();
    });

    it('handles an FS_NEW_FIDDLE command', (done) => {
      shallow(<Editors appState={store} />);

      ipcRendererManager.emit(IpcEvents.FS_NEW_FIDDLE, null);
      process.nextTick(() => {
        expect(window.ElectronFiddle.app.setValues).toHaveBeenCalled();
        done();
      });
    });

    it('handles the monaco editor option commands', () => {
      shallow(<Editors appState={store} />);

      ipcRendererManager.emit(IpcEvents.MONACO_TOGGLE_OPTION, null, 'wordWrap');

      expect(window.ElectronFiddle.editors.html!.updateOptions).toHaveBeenCalled();
      expect(window.ElectronFiddle.editors.renderer!.updateOptions).toHaveBeenCalled();
      expect(window.ElectronFiddle.editors.main!.updateOptions).toHaveBeenCalled();
    });
  });

  describe('loadMonaco()', () => {
    it('loads Monaco', (done) => {
      window.ElectronFiddle.app.monaco = null;

      shallow(<Editors appState={store} />);

      process.nextTick(() => {
        expect(window.ElectronFiddle.app.monaco).toEqual({ monaco: true});
        done();
      });
    });
  });
});
