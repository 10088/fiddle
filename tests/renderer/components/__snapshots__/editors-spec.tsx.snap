// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editors component renders 1`] = `
<div
  className="mosaic-blueprint-theme mosaic mosaic-drop-target"
>
  <div
    className="mosaic-zero-state bp3-non-ideal-state"
  >
    <div
      className="bp3-non-ideal-state-visual"
    >
      <span
        className="bp3-icon bp3-icon-applications"
        icon="applications"
      >
        <svg
          data-icon="applications"
          height={120}
          viewBox="0 0 20 20"
          width={120}
        >
          <desc>
            applications
          </desc>
          <path
            d="M15 5H1c-.55 0-1 .45-1 1v13c0 .55.45 1 1 1h14c.55 0 1-.45 1-1V6c0-.55-.45-1-1-1zm-1 13H2V8h12v10zM3.5 10h7c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-7c-.28 0-.5.22-.5.5s.22.5.5.5zm0 2h3c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-3c-.28 0-.5.22-.5.5s.22.5.5.5zm0 2h5c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-5c-.28 0-.5.22-.5.5s.22.5.5.5zM19 0H5c-.55 0-1 .45-1 1v3h2V3h12v10h-1v2h2c.55 0 1-.45 1-1V1c0-.55-.45-1-1-1z"
            fillRule="evenodd"
            key="0"
          />
        </svg>
      </span>
    </div>
    <h4
      className="bp3-heading"
    >
      No Windows Present
    </h4>
    <div />
  </div>
  <div
    className="drop-target-container"
  >
    <div
      className="drop-target top"
    />
    <div
      className="drop-target bottom"
    />
    <div
      className="drop-target left"
    />
    <div
      className="drop-target right"
    />
  </div>
</div>
`;

exports[`Editors component renders a toolbar 1`] = `
<div>
  <div>
    <h5>
      Main Process
    </h5>
  </div>
  <div />
  <div
    className="mosaic-controls"
  >
    <MaximizeButton
      appState={
        Object {
          "isSettingsShowing": false,
          "isTokenDialogShowing": false,
          "setWarningDialogTexts": [Function],
        }
      }
      id="main"
    />
    <RemoveButton
      appState={
        Object {
          "isSettingsShowing": false,
          "isTokenDialogShowing": false,
          "setWarningDialogTexts": [Function],
        }
      }
      id="main"
    />
  </div>
</div>
`;
